### PR TITLE
Reusing ReverseProxy

### PR DESCRIPTION
**Pull request by oxtoacart**
This requires a PR for servers to get deployed first so that they can handle multiple auth tokens - https://github.com/getlantern/http-proxy-lantern/pull/28

### CODE REVIEW

**Comment by myleshorton** [comment]
Looking good to me! Can you also update the Dockerfile to use 1.6 or is that not strictly necessary?

**Comment by oxtoacart** [comment]
@myleshorton Oh yeah, I forgot about the docker build.  Let me fix that.

**Comment by oxtoacart** [comment]
@myleshorton I've upgraded the go version in the Dockerfile to 1.6rc1.  It's able to build Windows and Linux binaries.

File: src/github.com/getlantern/flashlight/client/reverseproxy.go [hunk 1]
```diff
@@ -3,90 +3,49 @@ package client
...
 	}
 
-	rp := &httputil.ReverseProxy{
+	allAuthTokens := bal.AllAuthTokens()
+	return &httputil.ReverseProxy{
+		// We need to set the authentication tokens for all servers that we might
+		// connect to because we don't know which one the dialer will actually
+		// pick. We also need to strip out X-Forwarded-For that reverseproxy adds
+		// because it confuses the upstream servers with the additional 127.0.0.1
+		// field when upstream servers are trying to determin the client IP.
```
Comment (myleshorton) [inline src/github.com/getlantern/flashlight/client/reverseproxy.go]: nit: typo determin -> determine

**Comment by myleshorton** [comment]
This otherwise looks great to me @oxtoacart! Great stuff.

File: src/github.com/getlantern/flashlight/client/reverseproxy.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/client/reverseproxy.go]: Fixed.

**Comment by myleshorton** [comment]
Cool not sure what's up with Travis, but tests are passing for me locally with the exception of that detour test that's fails inconsistently but is unrelated. Merging!

**Comment by oxtoacart** [comment]
Thanks @myleshorton !

**Comment by myleshorton** [comment]
Thank you @oxtoacart! Great change.

**Comment by fffw** [comment]
@oxtoacart I saw strange behaviour with latest valencia code. With only one chained server in my config, I frequently get mimicked page, especially if the site is inaccessible. Known sites such as Twitter and FB never show that page. From the Lantern server, I can see lots of `No token provided from x.x.x.x:41601` message.

Is it expected with current http-proxy-lantern? It looks like in some cases, Lantern client never set the auth token in header.

**Comment by oxtoacart** [comment]
@fffw Can you send me your lantern config?

**Comment by myleshorton** [comment]
Is the multiple token checking code already deployed on the server?

**Comment by fffw** [comment]
Can't reproduce it now. Will keep experimenting.

**Comment by oxtoacart** [comment]
@myleshorton No, the multiple token checking code is not yet deployed.

**Comment by oxtoacart** [comment]
I wonder if it could be a timing in the initialization sequence ...

**Comment by oxtoacart** [comment]
@fffw If you run into the problem again, I wonder if this branch improves things for you:

https://github.com/getlantern/lantern/tree/initfix

**Comment by fffw** [comment]
Thanks @oxtoacart I'll try

**Comment by oxtoacart** [comment]
@fffw Something that may help you reproduce the problem is to remove the config files and start lantern again - the problem may only happen during the period when your first configuration is getting fetched.

**Pull request merged**